File: Matrix.cs
using System;

public static class Matrix
{

    public static double[] SumMatrix(double[] m, double[] m1, int rows, int cols)
    {

        double[] sum = new double[rows * cols];

        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++)
                sum[i * cols + j] = m[i * cols + j] + m1[i * cols + j];

        return sum;
    }


    public static void ProdScalar(double[] m, int rows, int cols, double scalar)
    {
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++)
                m[i * cols + j] *= scalar;
    }


    public static double[] RowsCols(double[] m, double[] m1, int rows, int cols, int cols1)
    {
        double[] prod = new double[rows * cols1];

        int i, k, j;
        double sum;

        for (i = 0; i < rows; i++)
        {
            for (k = 0; k < cols1; k++)
            {
                sum = 0;
                for (j = 0; j < cols; j++)
                    sum += (m[i * cols + j] * m1[j * cols1 + k]);

                prod[i * cols1 + k] = sum;
            }
        }

        return prod;
    }



    public static double[] Transpose(double[] m, int rows, int cols)
    {
        double[] transposed = new double[rows * cols];

        for (int i = 0; i < cols; i++)
            for (int j = 0; j < rows; j++)
                transposed[i * rows + j] = m[j * cols + i];

        return transposed;
    }


    public static double Determinant(double[] m, int rows, int cols)
    {
        double det;
        int i, j;

        if (rows == cols && rows == 1)
            return m[0];

        if (rows == 2 && rows == cols) // Se è una matrice 2x2
            return (m[0] * m[cols + 1]) - (m[1] * m[cols]);
        //return (*(m + (0 * cols) + 0) * *(m + (1 * cols) + 1)) - (*(m + (0 * cols) + 1) * *(m + (1 * cols) + 0));

        if (CheckIfDiag(m, rows, cols))
        {

            if (SupDiag(m, rows, cols) || InfDiag(m, rows, cols))
            {
                det = 1;

                for (i = 0; i < rows; i++)
                {
                    for (j = 0; j < cols; j++)
                    {
                        if (i == j)
                        {
                            det *= m[i * cols + j];
                            //*(m + (i * cols) + j);
                            break;
                        }
                    }
                }

                return det;
            }
        }

        if (ZeroRow(m, rows, cols) || ZeroCol(m, rows, rows))
            return 0;

        i = 0;
        det = 0;
        int k, p, copyRows = rows - 1;

        for (j = 0; j < cols; j++)
        {
            double[] tmp = new double[copyRows * copyRows];
            int t = 0;

            for (k = 1; k < rows; k++)
            {
                int y = 0;
                for (p = 0; p < cols; p++)
                {
                    if (p == j)
                        continue;
                    tmp[t * copyRows + y] = m[k * cols + p];
                    //*(tmp + (t * copyRows) + y) = *(m + (k * cols) + p);
                    y++;
                }
                t++;
            }

            det += Math.Pow(-1, (i + 1) + (j + 1)) * m[i * cols + j] * Determinant(tmp, copyRows, copyRows);
            Array.Clear(tmp, 0, tmp.Length);
        }
        return det;
    }


    private static bool CheckIfDiag(double[] m, int rows, int cols)
    {

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
                if (i == j && m[i * cols + j] == 0)
                    return false;
        }

        return true;
    }


    private static bool InfDiag(double[] m, int rows, int cols)
    {
        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols - 1; j++)
            {
                if (i == j)
                    break;

                if (i != j && m[i * cols + j] != 0)
                    return false;
            }
        }

        return true;
    }


    private static bool SupDiag(double[] m, int rows, int cols)
    {

        int halfj = cols / 2;

        for (int i = 0; i < rows - 1; i++)
        {
            for (int j = cols - 1; j > halfj; j--)

                if (m[i * cols + j] != 0)
                    return false;
        }
        return true;
    }


    private static bool ZeroRow(double[] m, int rows, int cols)
    {

        bool zeros = false;

        for (int i = 0; i < rows; i++)
        {
            zeros = true;
            for (int j = 0; j < cols; j++)
                if (m[i * cols + j] != 0)
                    zeros = false;

            if (zeros)
                break;
        }

        return zeros;
    }


    private static bool ZeroCol(double[] m, int rows, int cols)
    {
        bool zeros = false;

        for (int i = 0; i < cols; i++)
        {
            zeros = true;
            for (int j = 0; j < rows; j++)
                if (m[j * cols + i] != 0)
                    zeros = false;

            if (zeros)
                break;
        }

        return zeros;
    }


    public static double[] Algebrics(double[] m, int rows, int cols)
    {
        double[] algebric = new double[rows * cols];

        int i, j;
        double det = 0;
        int k, p, copyRows = rows - 1;

        for (i = 0; i < rows; i++)
        {
            for (j = 0; j < cols; j++)
            {
                double[] tmp = new double[copyRows * copyRows];

                int t = 0;

                for (k = 0; k < rows; k++)
                {
                    if (k == i)
                        continue;
                    int y = 0;

                    for (p = 0; p < cols; p++)
                    {

                        if (p == j)
                            continue;

                        tmp[t * copyRows + y] = m[k * cols + p];
                        y++;
                    }
                    t++;
                }
                det += Math.Pow(-1, (i + 1) + (j + 1)) * Determinant(tmp, copyRows, copyRows);
                algebric[i * cols + j] = det;
                det = 0;
                Array.Clear(tmp, 0, tmp.Length);
            }
        }

        return algebric;
    }

    public static void RiduzioneGaussiana(double[] matrix, int rows, int cols)
    {
        double result;

        for (int j = 0; j < cols; j++)
            for (int i = rows - 1; i >= 0; i--)
                for (int k = i - 1; k >= 0; k--)
                    if (matrix[i * cols + j] != 0 && matrix[k * cols + j] != 0 && ((j >= 1) ? matrix[k * cols + j - 1] == 0 : true))
                    {

                        result = -matrix[i * cols + j] / matrix[k * cols + j];
                        for (int h = j; h < cols; h++)
                            matrix[i * cols + h] = matrix[i * cols + h] + (matrix[k * cols + h] * result);

                        if (IsLadder(matrix, rows, cols))
                            return;
                    }
    }


    private static bool IsLadder(double[] matrix, int rows, int cols)
    {

        Pick(matrix, rows, cols);

        for (int j = 0; j < cols; j++)
            for (int i = 0; i < rows; i++)
                for (int k = i + 1; k < rows - 1; k++)
                    if (matrix[i * cols + j] != 0 && matrix[k * cols + j] != 0)
                        return false;
        return true;
    }


    private static void Pick(double[] matrix, int rows, int cols)
    {

        for (int i = 0; i < rows; i++)
            for (int j = i + 1; j < rows; j++)
                Compare(matrix, i, j, cols);
                //Compare(matrix[i], matrix[j], cols);

    }

    private static void Compare(double[] row, int i, int j, int cols)
    {

        for (; i < cols; i++)
            if (row[i] == 0 && row[i] != 0 && ((i >= 1) ? Zeros(row, i - 1) && Zeros(row, i - 1) : true))
                for (; j < cols; j++)
                    Utils.Swap<double>(ref row[j], ref row[j]);
    }


    private static bool Zeros(double[] row, int pos)
    {

        for (; pos >= 0; pos--)
            if (row[pos] != 0)
                return false;

        return true;
    }

    public static int Rank(double[] matrix, int righe, int colonne)
    {
        bool sw = false;
        int pivots = 0;

        for (int i = 0; i < righe; i++)
        {
            sw = true;
            for (int j = 0; j < colonne; j++)
            {
                if (matrix[i * colonne +  j] != 0 && sw)
                {
                    pivots++;
                    sw = false;
                }

            }
        }
        return pivots;
    }
}

File: matrix.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <math.h>
#include "nMatrix.h"
#include "zMatrix.h"
#include "utils.h"

//Prototypes
void nMain();
void zMain();
int choice();


int main(void) {
    
    char zN[3];

    printf("Do you want to do operations in Zn? Yes or No -->  ");
    scanf("%s", zN);
    printf("\n");

    if (zN[0] == 'y')
        zMain();
    else
        nMain();

    return 0;
}



void nMain() {
    int ans = 1;
    switch(choice()) {
        int rows, cols, rows1, cols1;
        double *m1, *m2, scalar;

        case 1:
            printf("\nGoing to calculate sum between two matrix.\n");

            get_dim(&rows, &cols); // MATRIX' S DIMENSIONS MUST BE EQUAL FOR SUMMING, NO NEED TO GET DOUBLE INPUT 

            m1 = (double *) malloc((rows * cols) * sizeof(double));
            m2 = (double *) malloc((rows * cols) * sizeof(double));

            check_malloc(m1);
            check_malloc(m2);
            
            printf("Insert A matrix\n");
            insert_matrix(m1, rows, cols);
            print_matrix(m1, rows, cols, &ans);

            while (!ans) {
                printf("Insert A matrix\n");
                insert_matrix(m1, rows, cols);
                print_matrix(m1, rows, cols, &ans);
            }

            printf("\nInsert B matrix\n");
            insert_matrix(m2,rows, cols);
            print_matrix(m2, rows, cols, &ans);

            while (!ans) {
                printf("Insert B matrix\n");
                insert_matrix(m2, rows, cols);
                print_matrix(m2, rows, cols, &ans);
            }

            printf("A + B = \n");
            print_result(sum_matrix(m1, m2, rows, cols), rows, cols);
            break; // ENDS CASE 1

        case 2:
            printf("\nGoing to calculate product for a scalar.\n");

            get_dim(&rows, &cols);
            m1 = (double *) malloc((rows * cols) * sizeof(double));

            check_malloc(m1);

            printf("Insert Matrix\n");
            insert_matrix(m1, rows, cols);
            print_matrix(m1, rows, cols, &ans);

     
            while (!ans) {
                printf("Insert Matrix\n");
                insert_matrix(m1, rows, cols);
                print_matrix(m1, rows, cols, &ans);
            }

            printf("Insert scalar: ");
            scanf("%lf", &scalar);

            printf("Matrix * %lf = \n", scalar);
            prod_scalar(m1,rows,cols, scalar);
            print_result(m1, rows, cols);
            break; // ENDS CASE 2

        case 3:
            printf("\nGoing to calculate product rows  * cols.\n");

            get_dim(&rows, &cols);
            m1 = (double *) malloc((rows * cols) * sizeof(double));
            check_malloc(m1);

            get_dim(&rows1, &cols1);

            if (cols != rows1) {
                printf("'A Matrix' s cols must be equal to B Matrix' s rows.\n");
                exit(-1);
            }

            m2 = (double *) malloc((rows1 * cols1) * sizeof(double));
            check_malloc(m2);
            
            printf("Insert A matrix\n");
            insert_matrix(m1, rows, cols);
            print_matrix(m1, rows, cols, &ans);

            while (!ans) {
                printf("Insert A matrix\n");
                insert_matrix(m1, rows, cols);
                print_matrix(m1, rows, cols, &ans);
            }


            printf("Insert B matrix\n");
            insert_matrix(m2, rows1, cols1);
            print_matrix(m2, rows1, cols1, &ans);

            while (!ans) {
                printf("Insert B matrix\n");
                insert_matrix(m2, rows1, cols1);
                print_matrix(m2, rows1, cols1, &ans);
            }

            print_result(rows_cols(m1,m2,rows,cols,rows1,cols1), rows, cols1);

            break; // ENDS CASE 3


        case 4:
            printf("\nGoing to calculate the transpose.\n");

            get_dim(&rows, &cols);
            m1 = (double *) malloc((rows * cols) * sizeof(double));

            check_malloc(m1);

            printf("Insert Matrix\n");
            insert_matrix(m1, rows, cols);
            print_matrix(m1, rows, cols, &ans);

     
            while (!ans) {
                printf("Insert Matrix\n");
                insert_matrix(m1, rows, cols);
                print_matrix(m1, rows, cols, &ans);
            }

            double *transposed = (double *) malloc((rows * cols) * sizeof(double));
            check_malloc(transposed);
            transposed = transpose(m1, rows, cols);

            print_result(transposed, rows, cols);

            break; // ENDS CASE 4

        case 5:
            printf("\nGoing to calculate determinant.\n");

            get_dim(&rows, &cols);

            if (rows != cols) {
                printf("Matrix must be square in order to calculate determinant.\n");
                exit(-3);
            }

            m1 = (double*) malloc((rows * cols) * sizeof(double));

            check_malloc(m1);

            printf("Insert A Matrix\n");
            insert_matrix(m1, rows, cols);
            print_matrix(m1, rows, cols, &ans);

     
            while (!ans) {
                printf("Insert A Matrix\n");
                insert_matrix(m1, rows, cols);
                print_matrix(m1, rows, cols, &ans);
            }

            double det = determinant(m1, rows, cols);
            double mod, intPart;

            mod = modf(det, &intPart);

            if (!mod)
                printf("Determinant: %.lf.\n", intPart);
            else
                printf("Determinant: %.2lf.\n", det);
            break;  // ENDS CASE 5

        case 6:
            printf("\nGoing to matrix's reverse.\n");

            get_dim(&rows, &cols);
            
            if (rows != cols) {
                printf("Matrix must be square in order to calculate reverse.\n");
                exit(-3);
            }

            m1 = (double*) malloc((rows * cols) * sizeof(double));
            check_malloc(m1);
            
            printf("Insert A Matrix\n");
            insert_matrix(m1, rows, cols);
            print_matrix(m1, rows, cols, &ans);

             while (!ans) {
                printf("Insert A Matrix\n");
                insert_matrix(m1, rows, cols);
                print_matrix(m1, rows, cols, &ans);
            }

            double det1 = determinant(m1, rows, cols);

            if (det1 == 0) {
                printf("Matrixes with null determinant cant be reversed.\n");
                break;
            }

            det1 = pow(det1, -1);

            double *algebric = (double *) malloc(rows * cols * sizeof(double));
            check_malloc(algebric);
            

            algebric = algebrics(m1, rows, cols);
            algebric = transpose(algebric, rows, cols);

            prod_scalar(algebric, rows, cols, det1);
            print_result(algebric, rows, cols);

            free(algebric);
            break; // ENDS CASE 6

        case 7:
            printf("\nGoing to matrix's reverse.\n");

            get_dim(&rows, &cols);
            
            if (rows != cols) {
                printf("Matrix must be square in order to calculate reverse.\n");
                exit(-3);
            }

            m1 = (double*) malloc((rows * cols) * sizeof(double));
            check_malloc(m1);
            
            printf("Insert A Matrix\n");
            insert_matrix(m1, rows, cols);
            print_matrix(m1, rows, cols, &ans);

             while (!ans) {
                printf("Insert A Matrix\n");
                insert_matrix(m1, rows, cols);
                print_matrix(m1, rows, cols, &ans);
            }

            double *algebric1 = (double *) malloc(rows * cols * sizeof(double));
            check_malloc(algebric1);

            algebric1 = algebrics(m1, rows, cols);
            print_result(algebric1, rows, cols);

            break; // ENDS CASE 7

        case 8:
            printf("\nGoing to calculate ladder matrix.\n");

            get_dimN(&rows, &cols);

            m1 = (double*) malloc((rows * cols) * sizeof(double));
            check_malloc(m1);
            
            printf("Insert A Matrix\n");
            insert_matrix(m1, rows, cols);
            print_matrix(m1, rows, cols, &ans);

            while (!ans) {
                printf("Insert A Matrix\n");
                insert_matrix(m1, rows, cols);
                print_matrix(m1, rows, cols, &ans);
            }

            third_operation(m1, rows, cols);
            print_result(m1, rows, cols);
            break; // ENDS CASE 8

            free(m1);
            if (choice() != 1 || choice() != 3)
                free(m2);
    }
}


void zMain() {

    int n;
    printf("Insert N, everything will be converted to the domain --> ");
    scanf("%d", &n);
    printf("\n");

    int ans = 1;
    switch(choice()) {
        int rows, cols, rows1, cols1;
        int *m1, *m2, scalar;

        case 1:
            printf("\nGoing to calculate sum between two matrix.\n");

            get_dimN(&rows, &cols); // MATRIX' S DIMENSIONS MUST BE EQUAL FOR SUMMING, NO NEED TO GET DOUBLE INPUT 

            m1 = (int *) malloc((rows * cols) * sizeof(int));
            m2 = (int *) malloc((rows * cols) * sizeof(int));

            check_mallocN(m1);
            check_mallocN(m2);
            
            printf("Insert A matrix\n");
            insert_matrixN(m1, rows, cols, n);
            print_matrixN(m1, rows, cols, &ans);

            while (!ans) {
                printf("Insert A matrix\n");
                insert_matrixN(m1, rows, cols, n);
                print_matrixN(m1, rows, cols, &ans);
            }

            printf("\nInsert B matrix\n");
            insert_matrixN(m2,rows, cols, n);
            print_matrixN(m2, rows, cols, &ans);

            while (!ans) {
                printf("Insert B matrix\n");
                insert_matrixN(m2, rows, cols, n);
                print_matrixN(m2, rows, cols, &ans);
            }

            printf("A + B = \n");
            print_resultN(sum_matrixN(m1, m2, rows, cols, n), rows, cols);
            break; // ENDS CASE 1

        case 2:
            printf("\nGoing to calculate product for a scalar.\n");

            get_dimN(&rows, &cols);
            m1 = (int *) malloc((rows * cols) * sizeof(int));

            check_mallocN(m1);

            printf("Insert Matrix\n");
            insert_matrixN(m1, rows, cols, n);
            print_matrixN(m1, rows, cols, &ans);

     
            while (!ans) {
                printf("Insert Matrix\n");
                insert_matrixN(m1, rows, cols, n);
                print_matrixN(m1, rows, cols, &ans);
            }

            printf("Insert scalar [0 --- %d]: ", n - 1);
            scanf("%d", &scalar);

            if (scalar < 0 || scalar > n - 1)
                scalar %= n;

            printf("Matrix * %d = \n", scalar);
            prod_scalarN(m1,rows,cols, scalar, n);
            print_resultN(m1, rows, cols);
            break; // ENDS CASE 2

        case 3:
            printf("\nGoing to calculate product rows  * cols.\n");

            get_dimN(&rows, &cols);
            m1 = (int *) malloc((rows * cols) * sizeof(int));
            check_mallocN(m1);

            get_dimN(&rows1, &cols1);

            if (cols != rows1) {
                printf("'A Matrix' s cols must be equal to B Matrix' s rows.\n");
                exit(-1);
            }

            m2 = (int *) malloc((rows1 * cols1) * sizeof(int));
            check_mallocN(m2);
            
            printf("Insert A matrix\n");
            insert_matrixN(m1, rows, cols, n);
            print_matrixN(m1, rows, cols, &ans);

            while (!ans) {
                printf("Insert A matrix\n");
                insert_matrixN(m1, rows, cols, n);
                print_matrixN(m1, rows, cols, &ans);
            }


            printf("Insert B matrix\n");
            insert_matrixN(m2, rows1, cols1, n);
            print_matrixN(m2, rows1, cols1, &ans);

            while (!ans) {
                printf("Insert B matrix\n");
                insert_matrixN(m2, rows1, cols1, n);
                print_matrixN(m2, rows1, cols1, &ans);
            }

            print_resultN(rows_colsN(m1,m2,rows,cols,rows1,cols1, n), rows, cols1);

            break; // ENDS CASE 3


        case 4:
            printf("\nGoing to calculate the transpose.\n");

            get_dimN(&rows, &cols);
            m1 = (int *) malloc((rows * cols) * sizeof(int));

            check_mallocN(m1);

            printf("Insert Matrix\n");
            insert_matrixN(m1, rows, cols, n);
            print_matrixN(m1, rows, cols, &ans);

     
            while (!ans) {
                printf("Insert Matrix\n");
                insert_matrixN(m1, rows, cols, n);
                print_matrixN(m1, rows, cols, &ans);
            }

            int *transposed = (int *) malloc((rows * cols) * sizeof(int));
            check_mallocN(transposed);
            transposed = transposeN(m1, rows, cols);

            print_resultN(transposed, rows, cols);

            break; // ENDS CASE 4

        case 5:
            printf("\nGoing to calculate determinant.\n");

            get_dimN(&rows, &cols);

            if (rows != cols) {
                printf("Matrix must be square in order to calculate determinant.\n");
                exit(-3);
            }

            m1 = (int*) malloc((rows * cols) * sizeof(int));

            check_mallocN(m1);

            printf("Insert A Matrix\n");
            insert_matrixN(m1, rows, cols, n);
            print_matrixN(m1, rows, cols, &ans);

     
            while (!ans) {
                printf("Insert A Matrix\n");
                insert_matrixN(m1, rows, cols, n);
                print_matrixN(m1, rows, cols, &ans);
            }

            int det = determinantN(m1, rows, cols, n);

            printf("Determinant: %d.\n", det);
            break;  // ENDS CASE 5

        case 6:
            printf("\nGoing to matrix's reverse.\n");

            get_dimN(&rows, &cols);
            
            if (rows != cols) {
                printf("Matrix must be square in order to calculate reverse.\n");
                exit(-3);
            }

            m1 = (int*) malloc((rows * cols) * sizeof(int));
            check_mallocN(m1);
            
            printf("Insert A Matrix\n");
            insert_matrixN(m1, rows, cols, n);
            print_matrixN(m1, rows, cols, &ans);

            while (!ans) {
                printf("Insert A Matrix\n");
                insert_matrixN(m1, rows, cols, n);
                print_matrixN(m1, rows, cols, &ans);
            }

            int det1 = determinantN(m1, rows, cols, n);
            det1 = classeDiEquivalenza(det1, n);
            int gcd = mcd(det1, n);

            if (det1 == 0 || gcd != 1) {
                printf("det = %d\n", det1);
                printf("Matrixes with null determinant or with unreversable determinant cant be reversed.\n");
                break;
            }

            int n2 = n;
            det1 /= gcd;
            n2 /= gcd;

            det1 = solve(det1, 1, n2);

            int *algebric = (int *) malloc(rows * cols * sizeof(int));
            check_mallocN(algebric);
            

            algebric = algebricsN(m1, rows, cols, n);
            algebric = transposeN(algebric, rows, cols);

            prod_scalarN(algebric, rows, cols, det1, n);
            print_resultN(algebric, rows, cols);

            free(algebric);
            break; // ENDS CASE 6

        case 7:
            printf("\nGoing to matrix's reverse.\n");

            get_dimN(&rows, &cols);
            
            if (rows != cols) {
                printf("Matrix must be square in order to calculate reverse.\n");
                exit(-3);
            }

            m1 = (int*) malloc((rows * cols) * sizeof(int));
            check_mallocN(m1);
            
            printf("Insert A Matrix\n");
            insert_matrixN(m1, rows, cols, n);
            print_matrixN(m1, rows, cols, &ans);

            while (!ans) {
                printf("Insert A Matrix\n");
                insert_matrixN(m1, rows, cols, n);
                print_matrixN(m1, rows, cols, &ans);
            }

            int *algebric1 = (int *) malloc(rows * cols * sizeof(int));
            check_mallocN(algebric1);

            algebric1 = algebricsN(m1, rows, cols, n);
            print_resultN(algebric1, rows, cols);

            break; // ENDS CASE 7

        case 8:
            printf("\nGoing to calculate ladder matrix.\n");

            get_dimN(&rows, &cols);

            m1 = (int*) malloc((rows * cols) * sizeof(int));
            check_mallocN(m1);
            
            printf("Insert A Matrix\n");
            insert_matrixN(m1, rows, cols, n);
            print_matrixN(m1, rows, cols, &ans);

            while (!ans) {
                printf("Insert A Matrix\n");
                insert_matrixN(m1, rows, cols, n);
                print_matrixN(m1, rows, cols, &ans);
            }

            break; // ENDS CASE 8

            free(m1);
            if (choice() != 1 || choice() != 3)
                free(m2);
    }
}


int choice() {
    // GUI MENU & INPUT VALIDATION

    int choice;

    printf("What operations do you want to do?\n");
    printf("Press 1 for sum\nPress 2 for product for a scalar\nPress 3 for product rows * cols\nPress 4 for transpose\nPress 5 for determinant\nPress 6 for reverse\nPress 7 for algebrics\nPress 8 for ladder -- >  ");
    scanf("%d", &choice);

    while (choice < 1 || choice > 8) {
        printf("Choice doesn't exist. Re-insert.\n");
        printf("Press 1 for sum\nPress 2 for product for a scalar\nPress 3 for product rows * cols\nPress 4 for transpose\nPress 5 for determinant\nPress 6 for reverse\nPress 7 for algebrics\nPress 8 for ladder -- >  ");
        scanf("%d", &choice);
    }

    return choice;
}